### PR TITLE
923 s3 bucket integration

### PR DESCRIPTION
**Pull request by JeremyQuartech**
This PR adds support for object storage in S3 or S3-like services such as BC Gov Object Store Service:

- Added object_Store.ts file to support s3 operations
- object_store relies on new environment variables to define the resource and auth (OBJECT_STORE_ACCESS_KEY_ID, OBJECT_STORE_SECRET_KEY_ID, OBJECT_STORE_BUCKET_NAME, OBJECT_STORE_URL)
- Artifact services are modified to interface with both main database and s3 storage
- A file must be provided on the api/artifact/create endpoint
- All api/artifact/ endpoints now return a signed url to the stored resource

### CODE REVIEW

File: src/utils/object_store.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+ * Local getter for retrieving the S3 object store URL.
+ *
+ * @returns {*} {string} The object store URL
+ * @throws {apiError} If OBJECT_STORE_URL environment variable is not defined
+ */
+const _getObjectStoreUrl = (): string => {
+  const url = process.env.OBJECT_STORE_URL;
+
+  if (!url) {
+    throw apiError.serverIssue();
```
Comment (MacQSL) [inline src/utils/object_store.ts:45]: Provide a message in the error thrown notifying no url found

File: src/utils/object_store.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+ * Local getter for retrieving the S3 object store bucket name.
+ *
+ * @returns {*} {string} The object store bucket name
+ * @throws {apiError} If OBJECT_STORE_BUCKET_NAME environment variable is not defined
+ */
+const _getObjectStoreBucketName = (): string => {
+  const bucketName = process.env.OBJECT_STORE_BUCKET_NAME;
+
+  if (!bucketName) {
+    throw apiError.serverIssue();
```
Comment (MacQSL) [inline src/utils/object_store.ts:61]: same here

File: src/utils/object_store.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+ *
+ * @returns {*} {AWS.S3} The S3 client
+ * @throws {apiError} If necessary environment variables are not defined
+ */
+const _getS3Client = (): AWS.S3 => {
+  const accessKeyId = process.env.OBJECT_STORE_ACCESS_KEY_ID;
+  const secretAccessKey = process.env.OBJECT_STORE_SECRET_KEY_ID;
+
+  if (!accessKeyId || !secretAccessKey) {
+    throw apiError.serverIssue();
```
Comment (MacQSL) [inline src/utils/object_store.ts:20]: here too

File: src/api/artifact/artifact.service.ts [hunk 4]
```diff
@@ -81,6 +103,17 @@ const deleteArtifact = async (artifact_id: string): Promise<artifact> => {
   });
 };
 
+/**
+ * * Maps over an array of artifacts and adds a signed URL to each one
+ * @param {artifact[]} artifacts
+ */
+const addSignedUrlToArtifacts = (artifacts: artifact[]): ArtifactResponse[] => {
```
Comment (MacQSL) [inline src/api/artifact/artifact.service.ts:110]: Could probably make this function accept (artifact | artifact[]) and return appropriate object.
Preventing having to call it like addSignedUrlToArtifacts([artifact])[0] etc

**Review by MacQSL — COMMENTED** [review]
Few slight changes

File: src/api/artifact/artifact.service.ts [hunk 5]
```diff
@@ -48,29 +57,42 @@ const updateArtifact = async (
   artifact_id: string,
   artifact_data: ArtifactUpdate
 ): Promise<artifact> => {
```
Comment (GrahamS-Quartech) [inline src/api/artifact/artifact.service.ts:59]: Should promise an ArtifactResponse, since it's returning from addSignedUrlToArtifacts

**Pull request merged**